### PR TITLE
Added bike theft data

### PR DESCRIPTION
**Pull request by thesun7385**

### CODE REVIEW

File: .github/workflows/run-update-sensor.yml [hunk 1]
```diff
@@ -165,3 +165,52 @@ jobs:
...
+      with:
+        version: ${{ env.UV_VERSION }}
+        enable-cache: true
+    - name: "Set up Python"
+      uses: actions/setup-python@ece7cb06caefa5fff74198d8649806c4678c61a1 # v6.3.0
+      with:
+        python-version-file: ".python-version"
+    - name: Install the project
+      run: uv sync --all-extras
+    - name: Run bike theft download script
```
Comment (tallcoleman) [inline .github/workflows/run-update-sensor.yml]: Ideally there should be one script that downloads the data and runs the ETL process

File: src/bikespace_data/bike_thefts/bicycle-thefts.geojson
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bicycle-thefts.geojson]: Output data should go on the `data` branch; ideally organized into an output directory such as `bicycle_thefts/`

File: src/bikespace_data/bike_thefts/stolen_bike_reports.json
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/stolen_bike_reports.json]: Output data should go on the `data` branch; ideally organized into an output directory such as `bicycle_thefts/`

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import urllib.request
+import os
+
+# Script to download the latest bike theft data from the City of Toronto's open data portal and save it as a GeoJSON file.
+URL = "https://ckan0.cf.opendata.inter.prod-toronto.ca/dataset/c7d34d9b-23d2-44fe-8b3b-cd82c8b38978/resource/e7fe6133-17d8-4a39-88af-352440dec684/download/bicycle-thefts%20-%204326.geojson"
+OUTPUT_DIR = os.path.dirname(os.path.abspath(__file__))
+OUTPUT_FILE = os.path.join(OUTPUT_DIR, "bicycle-thefts.geojson")
+
+def main():
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: `request_tod_gdf` from src/bikespace_data/resources/toronto_open_data.py would be a good fit for this task. It returns the data as a dict with two properties:

- `gdf` is the requested data, as a [geopandas](https://geopandas.org/en/stable/index.html) geodataframe
- `metadata` gives you the information from CKAN (the open data portal) about when the file was last updated, etc.

[validation]

The current behaviour where the original dataset is being saved right away is a smart move, because it allows you to go back and re-run the data transformation if needed, and it makes your results reproducible for others interested in understanding the data.

There is a utility function `save_geo_output` in src/bikespace_data/utilities/gdf_utils.py that makes it easy to save a geodataframe to disk and handles some common steps, such as ensuring the output file is properly formatted, and optionally saving an archive copy in parquet format.

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import json
+import re
+
+# This ETL script processes the bicycle thefts dataset, 
+# normalizing fields and filtering out certain premises types. 
+# It reads a GeoJSON file, transforms the data, and outputs 
+# a JSON file with the relevant information.
+
+# Mapping dictionaries for bike types and colors
+BIKE_TYPE_MAP = {
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: Is this taken from the documentation PDF linked in the description on https://open.toronto.ca/dataset/bicycle-thefts/ ? It looks like there are a couple options missing.

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    "PUR": "Purple",
+    "GRY": "Grey",
+    "SIL": "Silver",
+    "GLD": "Gold",
+    "BRN": "Brown",
+    "BGE": "Beige",
+    "PNK": "Pink",
+    "MRN": "Maroon",
+    "TAN": "Tan",
+    "SILRED": "Silver/Red",
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: Maybe you can map the 3-letter values only and then split the string and evaluate each colour in the combination if the length is 6?

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    "TRANSIT": "Transit",
+    "EDUCATIONAL": "Educational",
+}
+
+EXCLUDED_PREMISES = {"APARTMENT", "COMMERCIAL", "HOUSE"}
+
+
+def normalize_color(raw_color: str) -> str:
+    """Look up color code or attempt to clean raw value."""
+    if not raw_color or raw_color.strip().lower() == "none":
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: I think the `or` statement here might not get evaluated, since a string value of "None" would pass the truthiness test of `if not raw_color`.

Thinking ahead to how this might be rendered in the frontend, what about outputting an array of 0-2 colour strings? e.g.

- "None" -> `[]`
- "RED" -> `["Red"]`
- "BLKWHI" -> `["Black", "White"]`

These all appear to map to [named colours](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/named-color) so they could be rendered directly by the frontend with just the names.

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        return "Unknown"
+    upper = raw_color.strip().upper()
+    if upper in COLOR_MAP:
+        return COLOR_MAP[upper]
+    # Fallback: title-case the raw value
+    return raw_color.strip().title()
+
+
+def normalize_bike_type(raw_type: str) -> str:
+    if not raw_type or raw_type.strip().lower() == "none":
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: Same comment here about the `or` evaluation

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+# Function to transform data
+def convert_feature(feature: dict) -> dict:
+    props = feature["properties"]
+
+    # Coordinates — prefer top-level lat/long fields for precision
+    lat = props.get("LAT_WGS84")
+    lon = props.get("LONG_WGS84")
+
+    # Fallback to geometry coordinates if properties are missing
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: It looks like the Geometry coords are also missing in all the cases where the top-level lon lat are null

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    premises = props.get("PREMISES_TYPE", "").strip().upper()
+    if premises and premises.lower() != "none":
+        return LOCATION_TYPE_MAP.get(premises, premises.title() or "Unknown")
+    return "Unknown location"               # DIVISION is never touched
+
+# Function to transform data
+def convert_feature(feature: dict) -> dict:
+    props = feature["properties"]
+
+    # Coordinates — prefer top-level lat/long fields for precision
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: This is very interesting. How noticeable is the loss of precision?

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        "bikeType": normalize_bike_type(props.get("BIKE_TYPE", "")),
+        "color": normalize_color(props.get("BIKE_COLOUR", "")),
+        "description": build_description(props),
+        "status": STATUS_MAP.get(props.get("STATUS", "").upper(), "unknown"),
+        "latitude": lat,
+        "longitude": lon,
+    }
+
+
+def convert_geojson(input_path: str, output_path: str) -> None:
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: [geojson is natively handled by maplibre]
[can filter with a one-liner in geopandas]

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: [The organization of each step into individual functions is good]

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    # Convert to JSON
+    with open(output_path, "w", encoding="utf-8") as f:
+        json.dump(reports, f, indent=2, ensure_ascii=False)
+
+    # Display to the total records
+    print(f"Converted {len(reports)} records → {output_path}")
+
+
+if __name__ == "__main__":
+    convert_geojson("bicycle-thefts.geojson", "stolen_bike_reports.json")
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/bike-thefts-etl.py]: The paths need to be updated so that the files are saved in the `bike_theft` folder that the github action acts on.

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import urllib.request
+import os
+
+# Script to download the latest bike theft data from the City of Toronto's open data portal and save it as a GeoJSON file.
+URL = "https://ckan0.cf.opendata.inter.prod-toronto.ca/dataset/c7d34d9b-23d2-44fe-8b3b-cd82c8b38978/resource/e7fe6133-17d8-4a39-88af-352440dec684/download/bicycle-thefts%20-%204326.geojson"
+OUTPUT_DIR = os.path.dirname(os.path.abspath(__file__))
+OUTPUT_FILE = os.path.join(OUTPUT_DIR, "bicycle-thefts.geojson")
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: Re: the note below - to save a GeoDataFrame with `save_geo_output`, you can just use "bike_theft/bicycle-thefts.geojson" and shouldn't need any `os.path` wrapping.

If you have other situations in the future where you're saving a request into a file like this, check out [pathlib](https://docs.python.org/3/library/pathlib.html) in the standard library; it makes things like getting an absolute path and extending paths super easy.

**Review by tallcoleman — CHANGES_REQUESTED** [review]
Thanks for putting this together - I’m very excited to get a working bike theft map up on the site!

I’ve left some review comments for things to look at. My goal is to help get this dataset into the pipeline as soon as possible to enable the frontend to go live as soon as it’s ready. I’ve organized my comments into two sections: “review now” should be resolved in this PR, and “review before frontend is published” can be done in a future PR. Most other comments should be optional.

Review now

- The logic in `bike-thefts-etl.py` can be consolidated into `update_bike_thefts.py` so that you just have one script to call from the github action.
- The preferred way to work with downloaded data is with pandas/geopandas. There are several utility functions in the shared folders that help with common tasks like saving a geopandas dataframe to a geojson file or downloading a geojson file from the open data portal.
- The ideal output file format is `.geojson`; the MapLibre library that we use natively handles it well in an optimized way.
- The output files should be saved in the folder `bike_theft/` that the github action cherry picks over to the `data` branch on each run. Output files should generally not be saved to the `main` branch.
- One exception to that: I think it would be helpful to save the documentation PDFs that are currently linked from the open data portal since they serve as a reference for some of the dicts used in the ETL processing.
- A StatusManager instance should be added to track the metadata every time the data is updated. Lines 76-79, 93, and 133-139 in https://github.com/bikespace/parking-map-data/blob/main/src/bikespace_data/bicycle_network/update_cycling_network.py should give you a good example of usage.
- Once you have a StatusManager instance set up, you can use it to check if the source dataset has been updated and skip the rest of the ETL if it has not. This check/skip should be added since the dataset is only updated annually, but the script will run every day.

Review before frontend is published

- To prevent upstream changes to the data format from causing unanticpated script behaviour, I would reccommend adding validation for your source data using pandera (https://pandera.readthedocs.io/en/stable/). You can see a good example of this https://github.com/bikespace/parking-map-data/blob/main/src/bikespace_data/bicycle_network/update_cycling_network.py with how `cycling_network_schema` and `cycling_network_schema_optional` are used. This allows you to fail the processing script if e.g. an important column is removed or re-named in the source data and/or to log non-critical variances (e.g. a new location type is added).
- I would recommend trying to add at least one or two unit tests. Doing test cases for one of the functions like `normalize_color` would be a relatively easy way to start. Running unit tests with coverage will help you catch subtle bugs in your logic flow, ensure that edge cases are handled properly, and gives a lot more confidence when you’re updating or building on the codebase later.

One other thing to think about for the future, is that each point represents the closest intersection to the location of the bike theft. It would be interesting to line this dataset up against the City’s intersection dataset (https://open.toronto.ca/dataset/intersection-file-city-of-toronto/) to see if there are any intersections where there are no bike thefts reported nearby.

**Comment by tallcoleman** [comment]
Also FYI I ran our manual action to generate a preview: https://659fa6e7.bikespace-v2.pages.dev/stolen-history-map

File: src/bikespace_data/bike_thefts/bike-thefts-etl.py
Comment (thesun7385) [reply src/bikespace_data/bike_thefts/bike-thefts-etl.py]: TY, but looks like I did not save it in geojson haha, and I just realized that you guys have utility functions as well. Np, I will try to use them with the ETL process to generate the correct output.

**Review by thesun7385 — COMMENTED** [review]
@tallcoleman I just updated the function to save the file in geojson, and I also revised the job update bike stolen history data as well. Feel free to have a look!

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 12]
```diff
@@ -1,16 +1,132 @@
...
+import pandas as pd
+from bikespace_data.resources.toronto_open_data import request_tod_gdf
+from bikespace_data.utilities import StatusManager, save_geo_output
+
+# This script fetches bike stolen report then process ETL to generate a GeoJSON file for the BikeSpace app.
+# The data is sourced from the Toronto Open Data Portal.
+# Run command: python -m bikespace_data.bike_thefts.update_bike_thefts
+
+
+DATASET_NAME = "c7d34d9b-23d2-44fe-8b3b-cd82c8b38978"
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: `DATASET_NAME` should preferably be `"bicycle-thefts"` (name property) to make the error message more informative, but I guess the function works as expected if you supply the id instead!

File: src/bikespace_data/bike_thefts/statuses/bike_thefts_status.csv
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/statuses/bike_thefts_status.csv]: [check that this is getting output to the right path]

File: src/bikespace_data/bike_thefts/stolen_bike_reports.geojson
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/stolen_bike_reports.geojson]: [add note about removing data files before commit]

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 13]
```diff
@@ -1,16 +1,132 @@
...
+RESOURCE_ID = "e7fe6133-17d8-4a39-88af-352440dec684"
+
+OUTPUT_DIR = Path(__file__).parent
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: [run the script and check that things are landing in the right place; some of the stuff below is definitely wrong]

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 14]
```diff
@@ -1,16 +1,132 @@
...
+    "HOUSE": "House",
+    "TRANSIT": "Transit",
+    "EDUCATIONAL": "Educational",
+}
+
+EXCLUDED_PREMISES = {"APARTMENT", "COMMERCIAL"}
+
+# Function to map bike colors
+def normalize_color(raw_color) -> str:
+    if raw_color is None or pd.isna(raw_color) or str(raw_color).strip().lower() == "none":
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: `pd.isna(None)` returns `True`, so you might be able to simplify these guards by removing `if raw_color is None`

File: .github/workflows/run-update-sensor.yml [hunk 15]
```diff
@@ -167,50 +167,48 @@ jobs:
...
+      - name: Git setup
+        if: ${{ env.should_commit == 'True' }}
+        run: |
+          git config user.email ${{ secrets.BOT_EMAIL }}
+          git config user.name "tallcoleman-bot"
+          git fetch --depth=1
+      - name: Build commit on main (not pushed)
+        if: ${{ env.should_commit == 'True' }}
+        run: |
+          git add src/bikespace_data/bike_thefts
```
Comment (tallcoleman) [inline .github/workflows/run-update-sensor.yml]: this should just be `git add bike_thefts`

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 16]
```diff
@@ -1,16 +1,132 @@
...
+    return BIKE_TYPE_MAP.get(str(raw_type).strip().upper(), str(raw_type).strip().title())
+
+def main() -> None:
+    print("Fetching bike theft data from Toronto Open Data Portal...")
+    status_manager = StatusManager(
+        status_source=STATUS_SOURCE,
+        status_save=STATUS_PATH,
+    )
+    result = request_tod_gdf(DATASET_NAME, RESOURCE_ID)
+    gdf = result["gdf"]
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: Like the previous version, please save a copy of the "raw" open data in addition to your processed output. I would suggest saving to something like `bike_thefts/source_files/bicycle-thefts.geojson` or `bike_thefts/bicycle-thefts-source.geojson`

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 17]
```diff
@@ -1,16 +1,132 @@
...
+from bikespace_data.utilities import StatusManager, save_geo_output
+
+# This script fetches bike stolen report then process ETL to generate a GeoJSON file for the BikeSpace app.
+# The data is sourced from the Toronto Open Data Portal.
+# Run command: python -m bikespace_data.bike_thefts.update_bike_thefts
+
+
+DATASET_NAME = "c7d34d9b-23d2-44fe-8b3b-cd82c8b38978"
+RESOURCE_ID = "e7fe6133-17d8-4a39-88af-352440dec684"
+
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: For lines 16-22 below, a couple adjustments to make:

The data output folder that holds your source file, output file, and status file is separate from the `src/` path: it gets built on each script run as a throwaway commit in its own folder (in this case `bike_thefts/`), cherry picked to the `data` branch, and saved permanently there.

So I think you want `OUTPUT_DIR = "bike_thefts"` and then make sure to set the second string for `STATUS_SOURCE` to just `"bike_thefts/statuses/bike_thefts_status.csv"`

You will also want to pass any filepath constants into your functions as optional arguments, since this makes it super easy to do unit testing later with pytest's `tmp_path` fixture to give you a throwaway file directory, e.g.

```py
MY_CONSTANT = "folder_path/sub_folder/"
MY_FILENAME = "cooldata.geojson"

def my_function(save_here=MY_CONSTANT):
    ...
    save_geo_output(gdf, path=save_here, file_name=MY_FILENAME)
```

Then when you're testing, you can just call the function e.g. like this:

```py
my_function(tmp_path)
# assert expected output by checking what is in tmp_path directory
```

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 18]
```diff
@@ -1,16 +1,132 @@
...
+    gdf["location"] = gdf["PREMISES_TYPE"].apply(normalize_location)
+    gdf["bikeType"] = gdf["BIKE_TYPE"].apply(normalize_bike_type)
+    gdf["color"] = gdf["BIKE_COLOUR"].apply(normalize_color)
+    gdf["status"] = gdf["STATUS"].apply(
+        lambda x: STATUS_MAP.get(str(x).strip().upper(), "unknown")
+    )
+    gdf["date"] = gdf["OCC_DATE"].astype(str).str[:10]
+
+    # Save in GeoJSON format
+    save_geo_output(gdf, path=OUTPUT_DIR, file_name=OUTPUT_FILE)
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: Your output file has all the original columns (e.g. "BIKE_COLOUR") and your normalized ones (e.g. "color"). If you want to just save the normalized columns, I would suggest [selecting a subset](https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html) like this:

```py
# special GeoPandas "geometry" column needs to be included to ensure the output has coordinates
output_gdf = gdf[["location", "bikeType", "color", "status", "date", "geometry"]]
save_geo_output(output_gdf, ...
```

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 19]
```diff
@@ -1,16 +1,132 @@
...
+
+def main() -> None:
+    print("Fetching bike theft data from Toronto Open Data Portal...")
+    status_manager = StatusManager(
+        status_source=STATUS_SOURCE,
+        status_save=STATUS_PATH,
+    )
+    result = request_tod_gdf(DATASET_NAME, RESOURCE_ID)
+    gdf = result["gdf"]
+
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: Please add a check for whether the source data has been updated. The key parts would be:

Last modified date from open data portal:

```py
datetime.fromisoformat(result["metadata"]["last_modified"]).replace(tzinfo=timezone.utc)
```

Last modified date from status manager, which returns either a tz-aware `datetime` or `None`:
```py
sm.last_updated()
```

And then you can exit the function if the sm last modified is not `None` and the data portal last modified is not greater than the status manager last modified.

**Review by tallcoleman — CHANGES_REQUESTED** [review]
Thanks for making these updates! I've left some additional comments and once you've gone through those, we should be ready to merge.

I realized I should add better contributor documentation; I've opened [pr 49](https://github.com/bikespace/parking-map-data/pull/49) to do that and you can see the [draft documentation here](https://github.com/bikespace/parking-map-data/blob/7abc223b598ee081e45ca186f1bea8841d98c190/CONTRIBUTING.md). Take a look through and let me know if there's anything there that could be improved?

For the final PR, please make sure that it only includes the source files and does not include the `.geojson` or status manager file outputs.

It's a bit tricky during development because you can't `.gitignore` the outputs (otherwise the data update action wouldn't work). If you want to be able to run the script and see the output files during development without having git track them and without breaking the update action, you can [add the output folder to your local `.git/info/exclude`](https://stackoverflow.com/questions/1753070/how-do-i-configure-git-to-ignore-some-files-locally).

File: src/bikespace_data/bike_thefts/update_bike_thefts.py (hunk 19 above)
Comment (thesun7385) [reply src/bikespace_data/bike_thefts/update_bike_thefts.py]: This is the good one. I was previously wondering how did you guys check the update. TY!

**Review by thesun7385 — COMMENTED** [review]
@tallcoleman  Thank you for your prompt review. I just updated the file paths, git action, the save function. I also excluded unnecessary files that you mention earlier. Feel free to have a look!

File: src/bikespace_data/bike_thefts/source_files/bicycle-thefts_raw.geojson
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/source_files/bicycle-thefts_raw.geojson]: This output file should be removed

File: src/bikespace_data/bike_thefts/README.md
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/README.md]: Great addition!

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+    ):
+        print("Bike theft data has not been updated; exiting.")
+        return
+
+    gdf = result["gdf"]
+    
+    # Save a raw dataset in the sources_file for reference
+    save_geo_output(gdf, path=output_dir, file_name=output_source_file)
+
+    # Filter out excluded premises
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: (Optional) I noticed an issue that I've run into with other Toronto Open Data datasets - sometimes they store the geometry as a MultiPoint even when it's really just a Point. It can be fixed with the [explode function](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.explode.html):

```py
gdf = gdf.explode(index_parts=False)
```

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+# This script fetches bike stolen report then process ETL to generate a GeoJSON file for the BikeSpace app.
+# The data is sourced from the Toronto Open Data Portal.
+# Run command: python -m bikespace_data.bike_thefts.update_bike_thefts
+
+# Database and resource name
+DATASET_NAME = "bicycle-thefts"
+RESOURCE_ID = "e7fe6133-17d8-4a39-88af-352440dec684"
+
+# File Paths for output and source files
+OUTPUT_DIR = Path("bikespace_data/bike_thefts")
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: `OUTPUT_DIR` should just be `"bike_thefts"`

File: src/bikespace_data/bike_thefts/update_bike_thefts.py [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+        status_last_modified is not None
+        and portal_last_modified <= status_last_modified
+    ):
+        print("Bike theft data has not been updated; exiting.")
+        return
+
+    gdf = result["gdf"]
+    
+    # Save a raw dataset in the sources_file for reference
+    save_geo_output(gdf, path=output_dir, file_name=output_source_file)
```
Comment (tallcoleman) [inline src/bikespace_data/bike_thefts/update_bike_thefts.py]: You found a bug in `save_geo_output` - by default it creates the folder path for `output_dir` to prevent a "folder doesn't exist" error. However, it doesn't currently create a folder path if one is implied within `file_name`.

If you change line 31 as follows in `src/bikespace_data/utilities/gdf_utils.py` it should fix it:

```py
# current line 31
path.mkdir(exist_ok=True, parents=True)

# change it to this
(path / file_name).parent.mkdir(exist_ok=True, parents=True)
```

**Review by tallcoleman — CHANGES_REQUESTED** [review]
Thanks for making those updates! Nearly there, just two blocking issues that need to be resolved:

- fix`OUTPUT_DIR` and a bug you've helped us find with `save_geo_output`
- remove output file `src/bikespace_data/bike_thefts/source_files/bicycle-thefts_raw.geojson`

File: src/bikespace_data/bike_thefts/update_bike_thefts.py (hunk 22 above)
Comment (thesun7385) [reply src/bikespace_data/bike_thefts/update_bike_thefts.py]: TY! I just realized there is a condition the check the existing path.

**Review by thesun7385 — COMMENTED** [review]
@tallcoleman  As promised, the file path was fixed. Feel free to take a look. I did added a new function to build additional description.  Feel free to have a look!

**Review by tallcoleman — APPROVED** [review]
Thanks for your hard work! This is now approved and ready to merge.

Please squash merge and make sure to add a descriptive message to the merge commit.

File: .gitignore [hunk 23]
```diff
@@ -6,4 +6,11 @@ __pycache__/
 .ipynb_checkpoints
 .coverage*
 /.hypothesis
-.env
\ No newline at end of file
+.env
+
+# Generated bike theft data outputs
```
Comment (tallcoleman) [inline .gitignore]: Wait, actually you need to remove these two lines, otherwise the files won't get committed over to the `data` branch:

```
/bike_thefts/*.geojson
/bike_thefts/statuses/
```

If you have some of these in your working files, just delete or untrack them before committing.

**Review by tallcoleman — COMMENTED** [review]
One last thing to fix and then you're good to merge

File: .gitignore (hunk 23 above)
Comment (thesun7385) [reply .gitignore]: Great catch! I completely forgot to remove after correcting the dir paths.

**Comment by thesun7385** [comment]
### Bike theft dataset
Thank you @tallcoleman  for reviewing.
- Updated the ETL workflow (added the build‑description function and corrected file paths).
- Updated `geojson` output to follow project's standards.
- Updated the final file location for the `update-bike-theft-data` job in the `run-update-sensor.yml` 
- Fixed the final .gitignore by removing the bike_thefts paths from the root directory.
- Included `README.md` for additonal details
  
Merging Now!
